x/ecocredit/simulation/basket/operations.go

```diff
@@ -0,0 +1,422 @@
+	WeightTake   = 100
+)
+
+// ecocredit message types
+var (
+	TypeMsgCreate = basket.MsgCreate{}.Route()
+	TypeMsgPut    = basket.MsgPut{}.Route()
+	TypeMsgTake   = basket.MsgTake{}.Route()
+)
+
+func WeightedOperations(
+	appParams simtypes.AppParams, cdc codec.JSONCodec,
+	ak ecocredit.AccountKeeper, bk ecocredit.BankKeeper,
+	qryClient ecocredit.QueryClient, basketQryClient basket.QueryClient) simulation.WeightedOperations {
+
+	var (
+		weightMsgCreate int
+		weightMsgPut    int
+		weightMsgTake   int
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgCreate, &weightMsgCreate, nil,
+		func(_ *rand.Rand) {
+			weightMsgCreate = WeightCreate
+		},
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgPut, &weightMsgPut, nil,
+		func(_ *rand.Rand) {
+			weightMsgPut = WeightPut
+		},
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgTake, &weightMsgCreate, nil,
+		func(_ *rand.Rand) {
+			weightMsgTake = WeightTake
+		},
+	)
+
+	return simulation.WeightedOperations{
+		simulation.NewWeightedOperation(
+			weightMsgCreate,
+			SimulateMsgCreate(ak, bk, qryClient),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgPut,
+			SimulateMsgPut(ak, bk, qryClient, basketQryClient),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgTake,
+			SimulateMsgTake(ak, bk, qryClient, basketQryClient),
+		),
+	}
+}
+
+// SimulateMsgCreate generates a Basket/MsgCreate with random values.
+func SimulateMsgCreate(ak ecocredit.AccountKeeper, bk ecocredit.BankKeeper,
+	qryClient ecocredit.QueryClient) simtypes.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, sdkCtx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		curator, _ := simtypes.RandomAcc(r, accs)
+
+		ctx := regentypes.Context{Context: sdkCtx}
+		res, err := qryClient.Params(ctx, &ecocredit.QueryParamsRequest{})
+		if err != nil {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreate, err.Error()), nil, err
+		}
+
+		params := res.Params
+		spendable := bk.SpendableCoins(sdkCtx, curator.Address)
+		if spendable.IsAllLTE(params.BasketCreationFee) {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreate, "not enough balance"), nil, nil
+		}
+
+		classIds, err := randomClasses(r, ctx, qryClient)
+		if err != nil {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreate, err.Error()), nil, err
+		}
+
+		if len(classIds) == 0 {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreate, "no classes"), nil, nil
+		}
+
+		creditType, err := randomCreditType(r, ctx, qryClient)
+		if err != nil {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreate, err.Error()), nil, err
+		}
+
+		if creditType == nil {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreate, "credit type not found"), nil, nil
+		}
+
+		precision := int(creditType.Precision)
+		exponent := uint32(simtypes.RandIntBetween(r, precision, precision+5))
+
+		dateCriteria := randomDateCriteria(r)
+		dateCriteria = nil
```

Comment:
I'm getting this error if `dateCriteria` is not `nil`.

```
panic: Unregistered interface basket.isDateCriteria_Sum [recovered]
        panic: Unregistered interface basket.isDateCriteria_Sum [recovered]
        panic: Unregistered interface basket.isDateCriteria_Sum
```

Reply:
Are there any examples in the SDK where we use proto oneofs with amino? Maybe we do need to register it with amino...

Reply:
Can you try amino registering the interface and implementations and see if that works?

Reply:
> Are there any examples in the SDK where we use proto oneofs with amino? Maybe we do need to register it with amino...

Yes, authz/`StakeAuthorization`. I think that is also not working with amino.

Reply:
> Can you try amino registering the interface and implementations and see if that works?

```
        cdc.RegisterInterface((*isDateCriteria_Sum)(nil), nil)
	cdc.RegisterConcrete(&DateCriteria_MinStartDate{}, "regen.basket/DateCriteria_MinStartDate", nil)
	cdc.RegisterConcrete(&DateCriteria_StartDateWindow{}, "regen.basket/DateCriteria_StartDateWindow", nil)
```

Tried. Still getting same error.

Reply:
Well that's not good...